x/merkledb/db.go

```diff
@@ -65,8 +65,64 @@ type Config struct {
 	Tracer trace.Tracer
 }
 
-// Database can only be edited by committing changes from a trieView.
-type Database struct {
+type MerkleDB interface {
+	database.Database
+	Trie
+
+	// GetChangeProof returns a proof for a subset of the key/value changes in key range
+	// [start, end] that occurred between [startRootID] and [endRootID].
+	// Returns at most [maxLength] key/value pairs.
+	GetChangeProof(
+		ctx context.Context,
+		startRootID ids.ID,
+		endRootID ids.ID,
+		start []byte,
+		end []byte,
+		maxLength int,
+	) (*ChangeProof, error)
+
+	// Returns nil iff all of the following hold:
+	//   - [start] <= [end].
+	//   - [proof] is non-empty iff [proof.HadRootsInHistory].
+	//   - All keys in [proof.KeyValues] and [proof.DeletedKeys] are in [start, end].
+	//     If [start] is empty, all keys are considered > [start].
+	//     If [end] is empty, all keys are considered < [end].
+	//   - [proof.KeyValues] and [proof.DeletedKeys] are sorted in order of increasing key.
+	//   - [proof.StartProof] and [proof.EndProof] are well-formed.
+	//   - When the keys in [proof.KeyValues] are added to [db] and the keys in [proof.DeletedKeys]
+	//     are removed from [db], the root ID of [db] is [expectedEndRootID].
+	//
+	// Assumes [db.lock] isn't held.
+	// This is defined on Database instead of ChangeProof because it accesses
+	// database internals.
+	VerifyChangeProof(
+		ctx context.Context,
+		proof *ChangeProof,
+		start []byte,
+		end []byte,
+		expectedEndRootID ids.ID,
+	) error
+
+	// CommitChangeProof commits the key/value pairs within the [proof] to the db.
+	CommitChangeProof(ctx context.Context, proof *ChangeProof) error
+
+	// GetRangeProofAtRoot returns a proof for the key/value pairs in this trie within the range
+	// [start, end] when the root of the trie was [rootID].
+	GetRangeProofAtRoot(
```

Comment:
We could remove this method and just have `GetRangeProof` take in the rootID (i.e. keep this signature) but I'll leave that for later